presto-elasticsearch/src/main/java/io/prestosql/elasticsearch/decoders/IpAddressDecoder.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.elasticsearch.decoders;
+
+import com.google.common.net.InetAddresses;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import io.prestosql.spi.PrestoException;
+import io.prestosql.spi.block.BlockBuilder;
+import io.prestosql.spi.type.Type;
+import org.elasticsearch.search.SearchHit;
+
+import java.util.function.Supplier;
+
+import static io.airlift.slice.Slices.wrappedBuffer;
+import static io.prestosql.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.prestosql.spi.StandardErrorCode.INVALID_CAST_ARGUMENT;
+import static io.prestosql.spi.StandardErrorCode.TYPE_MISMATCH;
+import static java.lang.String.format;
+import static java.lang.System.arraycopy;
+import static java.util.Objects.requireNonNull;
+
+public class IpAddressDecoder
+        implements Decoder
+{
+    private final String path;
+    private final Type ipAddressType;
+
+    public IpAddressDecoder(String path, Type type)
+    {
+        this.path = requireNonNull(path, "path is null");
+        this.ipAddressType = requireNonNull(type, "type is null");
+    }
+
+    @Override
+    public void decode(SearchHit hit, Supplier<Object> getter, BlockBuilder output)
+    {
+        Object value = getter.get();
+        if (value == null) {
+            output.appendNull();
+        }
+        else if (value instanceof String) {
+            String address = (String) value;
+            Slice slice = castToIpAddress(Slices.utf8Slice(address));
+            ipAddressType.writeSlice(output, slice);
+        }
+        else {
+            throw new PrestoException(TYPE_MISMATCH, format("Expected a string value for field '%s' of type IP: %s [%s]", path, value, value.getClass().getSimpleName()));
+        }
+    }
+
+    /**
+      *this is a copy of IpAddressOperators.castFromVarcharToIpAddress method
+      */
```

Comment:
Nit: use a regular comment, not javadoc:

```suggestion
    // This is a copy of IpAddressOperators.castFromVarcharToIpAddress method
```

Reply:
done